### PR TITLE
Ian UID2-1677 services and service links

### PR DESCRIPTION
**Pull request by Ian-Nara**
[UID2-1677](https://atlassian.thetradedesk.com/jira/browse/UID2-1677)

requires the shared branch ian-UID2-1677-service-service-links

### CODE REVIEW

File: src/main/java/com/uid2/admin/Main.java [hunk 1]
```diff
@@ -162,6 +160,36 @@ public void run() {
...
             }
 
+            CloudPath serviceMetadataPath = new CloudPath(config.getString(Const.Config.ServiceMetadataPathProp));
+            GlobalScope serviceGlobalScope= new GlobalScope(serviceMetadataPath);
+            RotatingServiceStore serviceProvider = new RotatingServiceStore(cloudStorage, serviceGlobalScope);
+            ServiceStoreWriter serviceStoreWriter = new ServiceStoreWriter(serviceProvider, fileManager, jsonWriter, versionGenerator, clock, serviceGlobalScope);
+            try {
+                serviceProvider.loadContent();
+            } catch (CloudStorageException e) {
+                if(e.getMessage().contains("The specified key does not exist")) {
```
Comment (thomasm-ttd) [inline src/main/java/com/uid2/admin/Main.java:170]: This exception message looks like it is related to key sets, not services. Is this correct?

File: src/main/java/com/uid2/admin/Main.java [hunk 2]
```diff
@@ -162,6 +160,36 @@ public void run() {
...
+            }
+
+            CloudPath serviceLinkMetadataPath = new CloudPath(config.getString(Const.Config.ServiceLinkMetadataPathProp));
+            GlobalScope serviceLinkGlobalScope= new GlobalScope(serviceLinkMetadataPath);
+            RotatingServiceLinkStore serviceLinkProvider = new RotatingServiceLinkStore(cloudStorage, serviceLinkGlobalScope);
+            ServiceLinkStoreWriter serviceLinkStoreWriter = new ServiceLinkStoreWriter(serviceLinkProvider, fileManager, jsonWriter, versionGenerator, clock, serviceLinkGlobalScope);
+            try {
+                serviceLinkProvider.loadContent();
+            } catch (CloudStorageException e) {
+                if(e.getMessage().contains("The specified key does not exist")) {
```
Comment (thomasm-ttd) [inline src/main/java/com/uid2/admin/Main.java:185]: Same issue here - looks related to key sets, not service links

**Review by thomasm-ttd — COMMENTED** [review]
Good and comprehensive tests

File: src/main/java/com/uid2/admin/Main.java (hunk 1 above)
Comment (Ian-Nara) [reply src/main/java/com/uid2/admin/Main.java:170]: key refers to json key in the metadata. in this case the key would be either location, services, or service_links

**Review by thomasm-ttd — APPROVED** [review]

**Pull request merged**